lib/forwarder/handler.go

```diff
@@ -0,0 +1,210 @@
+// Handle the connection if a reservation can be obtained for the ClientID.
+// A ClientID is expected to be found in the context.
+type RateLimitingHandler struct {
+	Logger   slog.Logger
+	Reserver ClientReserver
+	Inner    Handler
+}
+
+func (h *RateLimitingHandler) Handle(ctx context.Context, conn AuthenticatedConn) {
+	clientID, ok := ClientIDFromContext(ctx)
+	if !ok {
+		h.Logger.Error(&slog.LogRecord{Msg: "RateLimitingHandler: Failed to get ClientID from context"})
+		return
+	}
+
+	// Clients are subject to rate-limiting.
+	err := h.Reserver.TryReserve(ctx, clientID)
+	if err != nil {
+		switch err {
+		// TODO: refactor to break dep on package lib/limiter
+		case limiter.MaxReservationsExceeded:
+			h.Logger.Warn(&slog.LogRecord{Msg: "RateLimitingHandler: Client rate limited", ClientID: &clientID})
+		default:
+			h.Logger.Error(&slog.LogRecord{Msg: "RateLimitingHandler: TryReserve error", ClientID: &clientID, Error: err})
+		}
+		return
+	}
+	defer func() {
+		err := h.Reserver.ReleaseReservation(ctx, clientID)
+		if err != nil {
+			h.Logger.Error(&slog.LogRecord{Msg: "RateLimitingHandler: ReleaseReservation error", ClientID: &clientID, Error: err})
+		}
+	}()
+
+	h.Inner.Handle(ctx, conn)
+}
+
+var _ Handler = (*RateLimitingHandler)(nil) // type check
+
+// AuthorizedUpstreamsHandler is a handler that determines which upstreams
+// the client connection is authorized to forward to. If the client is
+// authorized to connect to one or more upstreams, an UpstreamSet is stored
+// in the child context passed to the Inner Handler, and can be extracted
+// with UpstreamsFromContext.
+type AuthorizedUpstreamsHandler struct {
+	Logger     slog.Logger
+	Authorizer Authorizer
+	Inner      Handler
+}
+
+func (h *AuthorizedUpstreamsHandler) Handle(ctx context.Context, conn AuthenticatedConn) {
+	clientID, ok := ClientIDFromContext(ctx)
+	if !ok {
+		h.Logger.Error(&slog.LogRecord{Msg: "AuthorizedUpstreamsHandler: Failed to get ClientID from context"})
+		return
+	}
+
+	// Clients are only authorized to forward to certain upstreams.
+	authzUpstreams, err := h.Authorizer.AuthorizedUpstreams(ctx, clientID)
+	if err != nil {
+		h.Logger.Error(&slog.LogRecord{Msg: "AuthorizedUpstreamsHandler: AuthorizedUpstreams error", ClientID: &clientID, Error: err})
+		return
+	}
+	if len(authzUpstreams) == 0 {
+		h.Logger.Warn(&slog.LogRecord{Msg: "Client not authorized for forwarding", ClientID: &clientID, Error: err})
+		return
+	}
+
+	childCtx := NewContextWithUpstreams(ctx, authzUpstreams)
+
+	h.Inner.Handle(childCtx, conn)
+}
+
+var _ Handler = (*AuthorizedUpstreamsHandler)(nil) // type check
+
+// ForwardingHandler is the terminal handler that dials the best upstream to
+// serve the client connection, then forwards the client connection to that upstream.
+// It expects to find clientID and upstreams (the set of candidate upstreams to
+// consider forwarding to) in the given context.
+type ForwardingHandler struct {
```

Comment:
My overall design feedback would be to simplify handler types. We don't seem to need the composability, so my 2c is that it would be much simpler to figure out what the code does with less handlers.

Reply:
If you have a few minutes, I would be interested to hear your thoughts on the structure prior to the proliferation of "handlers" -- the first version I wrote had everything happening in a single Handle method of a Server type:

https://github.com/fcostin/tcplb/pull/11/commits/b10ad8fedc9a969b6c19da93cae7e2fa37f5f65b#diff-bb52046051a795d2837e24c92a0598f1027aa7c59b1bc224ba7b65f7b776afafL110

The previous code structure isn't ideal, but perhaps it is more readable in the direction you're suggesting ( ? ) -- there's a couple of pages of code in a large method that you can read in one hit, without having to learn and hold a bunch of different types and relationships in your mind.

The main prompt for me to switch to the current structure was thinking about testing - the composability allows each bit to be tested in isolation.  I agree the composability doesn't give any functional advantages (& if anything will make things run slower)

Reply:
Yep, I find that commit much easier to read/reason about.

Re APIs, composability and testing, it's all a bit intertwined. Composability can be desirable, but if there's isn't a clear purpose for it I'd rather have a simpler public API surface and test from it. The program does the same number of things, so that's what we'll be testing regardless, if it makes sense.